### PR TITLE
allow multiple windowing systems on linux

### PR DESCRIPTION
**Pull request by lrusak**
It's been a thorn in the side for a while to provide 6 linux binaries per architecture (which we don't even have CI for them all).

This will reduce the burden down to 2 linux binaries per architecture (one for each render system).

I've done this pretty quickly but to sum up the changes:
 - add a windowing system registration system (based on which windowing systems are built).
 - allow cmake to build multiple windowing platforms together (1, 2, or all 3).
 - change to use `APP_RENDER_SYSTEM` for linux builds
 - remove platform specific settings and move them into the linux.xml. The reason for this is that the platform settings are needed during windowing init. It's also not possible to remove settings once they are loaded.
 - add a command switch `--windowing=` for linux to specify the windowing to use (kodi will try all the window systems in order if not specified).
 - the app binary suffix will only be applied if one window system is being built

Areas that need improvement:
 - The cmake change to `CORE_PLATFORM_NAME_LC` now makes it a list. This is a bit ugly but I didn't want to add another cmake variable.
 - The `CEGLContextUtils` changes are a bit ugly but the old types are just typedefs anyways.
 - The `CPlatformLinux` changes look a bit ugly with all the ifdefs but I don't see any other way to keep the build flexible.

to build all windowing systems you need to use:
```
cmake -DCORE_PLATFORM_NAME="x11 wayland gbm" -DAPP_RENDER_SYSTEM="gles" ..
```

The help text will show the windowing systems that are available (built in)
```
$ ./kodi.bin --help
Usage: kodi [OPTION]... [FILE]...

Arguments:
  --windowing=<system>  Select which windowing method to use.
                                Available window systems are: wayland x11 gbm
```

### CODE REVIEW

**Comment by wsnipex** [comment]
works great for me on x11 and gbm, didn't have a change to test wayland yet.

issues:
- kodi.bin --windowing=wayland|x11|gbm on an "incorrect windowing env" crashes or locks keyboard input
  i.e. when on X11 kodi.bin --windowing=wayland crashes
- cmake output doesn't look nice(my original cmake patch fixed this):
```-- #---- CONFIGURATION ----#
-- Platforms: X11
-- WAYLAND
-- GBM
```

**Comment by wsnipex** [comment]
I thought so as well, but it is not fixed

**Review by yol — APPROVED** [review]
Really nice :sunglasses:

File: xbmc/windowing/X11/VideoSyncOML.h [hunk 1]
```diff
@@ -28,6 +36,9 @@ class CVideoSyncOML : public CVideoSync, IDispResource
 
 private:
   std::atomic_bool m_abort;
-  CWinSystemX11GLContext &m_winSystem;
+  CWinSystemX11GLContext* m_winSystem;
```
Comment (garbear) [inline xbmc/windowing/X11/VideoSyncOML.h:39]: I prefer `CWinSystemX11GLContext& m_winSystem` for safety against pointer dereferences. If polymorphism is what you're after, both `*` and `&` hit the vtable. If `*` is preferred, then deference safety can be increased with:

```c++
CWinSystemX11GLContext* const m_winSystem;
```

**Comment by lrusak** [comment]
rebased after the latest set of merges

File: docs/README.Linux.md [hunk 2]
```diff
@@ -141,25 +141,33 @@ cd $HOME/kodi-build
...
 
 Or configure build for GBM:
 ```
-cmake ../kodi -DCMAKE_INSTALL_PREFIX=/usr/local -DCORE_PLATFORM_NAME=gbm -DGBM_RENDER_SYSTEM=gles
+cmake ../kodi -DCMAKE_INSTALL_PREFIX=/usr/local -DCORE_PLATFORM_NAME=gbm -DAPP_RENDER_SYSTEM=gles
 ```
 
 **NOTE:** You can use `gl` instead of `gles` if you want to build with `GL`.
 
+Or configure build with any combination of the three:
```
Comment (wsnipex) [inline docs/README.Linux.md:163]:
```suggestion
Or configure build with any combination of the three (default is "x11 wayland gbm"):
```

**Comment by lrusak** [comment]
updated with suggested changes and fixed the GL build

**Review by wsnipex — APPROVED** [review]
great stuff

File: xbmc/CompileInfo.h [hunk 3]
```diff
@@ -30,5 +30,6 @@ class CCompileInfo
   static const char* GetCopyrightYears();
   static std::string GetBuildDate();
   static const char* GetVersionCode();
+  static const std::vector<std::string> GetAvailableWindowSystems();
```
Comment (yol) [inline xbmc/CompileInfo.h:33]:
```suggestion
  static std::vector<std::string> GetAvailableWindowSystems();
```

Don't return `const`

File: xbmc/CompileInfo.cpp.in [hunk 4]
```diff
@@ -100,3 +100,8 @@ std::vector<ADDON::RepoInfo> CCompileInfo::LoadOfficialRepoInfos()
 
   return officialRepoInfos;
 }
+
+const std::vector<std::string> CCompileInfo::GetAvailableWindowSystems()
```
Comment (yol) [inline xbmc/CompileInfo.cpp.in:104]:
```suggestion
std::vector<std::string> CCompileInfo::GetAvailableWindowSystems()
```

Don't return `const`

File: xbmc/windowing/WindowSystemFactory.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      std::find_if(m_windowSystems.begin(), m_windowSystems.end(),
+                   [&name](auto& windowSystem) { return windowSystem.first == name; });
+  if (windowSystem != m_windowSystems.end())
+    return windowSystem->second();
+
+  return nullptr;
+}
+
+void CWindowSystemFactory::RegisterWindowSystem(
+    std::function<std::unique_ptr<CWinSystemBase>()> createFunction, const std::string name)
```
Comment (yol) [inline xbmc/windowing/WindowSystemFactory.cpp:39]:
```suggestion
    std::function<std::unique_ptr<CWinSystemBase>()> createFunction, const std::string& name)
```

File: xbmc/windowing/WindowSystemFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+std::list<std::string> CWindowSystemFactory::GetWindowSystems()
+{
+  std::list<std::string> available;
+  for (const auto& windowSystem : m_windowSystems)
+    available.emplace_back(windowSystem.first);
+
+  return available;
+}
+
+std::unique_ptr<CWinSystemBase> CWindowSystemFactory::CreateWindowSystem(std::string name)
```
Comment (yol) [inline xbmc/windowing/WindowSystemFactory.cpp:27]:
```suggestion
std::unique_ptr<CWinSystemBase> CWindowSystemFactory::CreateWindowSystem(const std::string& name)
```

File: xbmc/windowing/WindowSystemFactory.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace KODI
+{
+namespace WINDOWING
+{
+
+class CWindowSystemFactory
+{
+public:
+  static std::unique_ptr<CWinSystemBase> CreateWindowSystem(std::string name);
```
Comment (yol) [inline xbmc/windowing/WindowSystemFactory.h:26]:
```suggestion
  static std::unique_ptr<CWinSystemBase> CreateWindowSystem(const std::string& name);
```

File: xbmc/windowing/WindowSystemFactory.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  static std::list<std::string> GetWindowSystems();
+  static void RegisterWindowSystem(std::function<std::unique_ptr<CWinSystemBase>()> createFunction,
+                                   std::string name = "default");
```
Comment (yol) [inline xbmc/windowing/WindowSystemFactory.h:29]:
```suggestion
                                   const std::string& name = "default");
```

**Review by yol — CHANGES_REQUESTED** [review]

**Comment by lrusak** [comment]
Probably worth a review now from all involed @wsnipex @yol @a1rwulf @fritsch

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by wsnipex** [comment]
due to GBM changes before this PR I can't test it on my nvidia prime notebook anymore, cause it only tries to init /dev/drm/card1, which is nouveau instead of /dev/drm/card0
Also reported on slack some days ago

**Comment by lrusak** [comment]
I'll have to adjust jenkins to get this to build properly. Any other thoughts?

**Comment by lrusak** [comment]
I've kicked off some test builds if we could have some people test windows, osx, android, etc https://jenkins.kodi.tv/view/Automation/job/BuildMulti-PR-Manually/399/

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
merge? :smile_cat:

**Comment by lrusak** [comment]
Thanks all involved in this one :) it took a bit of effort after all but will be nice to have for V19

**Comment by basilgello** [comment]
Anyone experiences inability to shut PC down / reboot after this fix? The "Selected logind/UPower as PowerSyscall" message is no more present in kodi.log, and the "Power Settings" menu contains only "Exit" regardless of polkit overrides. My system is Debian 10.

**Comment by howie-f** [comment]
i experience the same like @basilgello on debian 10. no power down / restart on the menu. not sure if it is this pr but presumably

**Comment by fritsch** [comment]
Open an issue, please - with a debuglog. For now two times Debian 10 affected, that needs tracking.

**Comment by wsnipex** [comment]
it's no debian only, happens on ubuntu as well. From a quick glance, it seems as if the power settings are not loaded or too late.

**Comment by stsichler** [comment]
Hi Irusak,
thank you for that feature!

How do think about also allowing the "app render system" to be defined by a command line switch in the same way "windowing"  can now  be set?

I have the problem here that the standard Debian package of Kodi for arm(/64) is set to GL rendering, but it would be very nice if it used GLES instead on platforms like RasPi or Odroid.
So this may be another step in order to have a unified package for all Linux platforms, now that DRMPRIME allows hardware acceleration out-of-box without patched sources (but currently requiring GLES on these platforms to work).

So, the `APP_RENDER_SYSTEM` build variable could be set to `"gl gles"` to pull in both dependencies and selection could be done at runtime.
But I'm not sure whether it would be better to merge that into the "--windowing" switch (like `--windowing=gbm-gles`) or provide a seperate switch "--rendering".

What do you think?

**Comment by stsichler** [comment]
Hm, you're right. At least [xbmc/system_gl.h](https://github.com/xbmc/xbmc/blob/9f6c9b15c3ed911a5673a531e3968683f3a3940d/xbmc/system_gl.h#L4) that currently includes either GL oder GLES mutually exclusive seems to cause a little include mess then. Not sure whether it is possible to change that to support both GL and GLES at the same time.

Another point to think about are surely the resulting blown up binary and package dependencies: 
The current `kodi.bin` binary and hence the Debian package already depends on _all_ of `libgbm1`, `libwayland*` _and_ `libx11` at the same time. By default, it also depends on `libgl1`. If we now add `libgles2`, it is further blowing up. This may not be desireable, especially on embedded platforms.
So, if we add such feature, I think it may be useful to make these package dependencies alternatives to each other, ie. the package depending on  `( libgbm1 | libwayland* | libx11)` and `(libgl1 | libgles2 )`. Of course, that would only make sense if the code would load these libraries at runtime using `dlopen()' instead of statically linking to them at build time.

**Comment by lrusak** [comment]
@stsichler see this PR -> https://github.com/xbmc/xbmc/pull/20604

That PR alone doesn't allow rendering system selection but it does the ground work for it.

The rendering system selection is implemented in this branch -> https://github.com/lrusak/xbmc/commits/gl-gles-prototypes-pr-2

As you can see it's a lot of work. Ultimately I think we decided against it at last devcon because there was a lot of overhead (generated headers) with not much gain (switching render systems isn't very practical and only benefits Linux).

So as much as I would like to do it I don't think it's something we will push for it.

It was a fun technical accomplishment though.

**Comment by stsichler** [comment]
What I shame! But I now do understand the decision against it. :-/

Hm, perhaps it makes more sense then to take a step back here and again offer different binaries and/or packages for GL and GLES on Linux, as was previously the case with X11, GBM and Wayland.
I think that could be a good idea, since with the introduction of DRMPRIME it seems usually no longer necessary to build a patched binary to get hardware acceleration.
But this is surely the same question as it was with kodi-x11, kodi-gbm and kodi-wayland.

**Comment by basilgello** [comment]
@stsichler I'd not call it a shame. In Debian, for example, I can pick whatever patchset I want (and rebase it, but it is another story!) If there is really strong demand from Arm users, every decision can be revisited.

**Comment by lrusak** [comment]
What feature do you think is missing from GLES compared to GL and vice versa?

**Comment by basilgello** [comment]
@lrusak GL does not work with HW acceleration on NanoPi-R5C and the platform I could not identify from this PR: https://salsa.debian.org/multimedia-team/kodi-media-center/kodi/-/merge_requests/10

**Comment by lrusak** [comment]
> @lrusak GL does not work with HW acceleration on NanoPi-R5C and the platform I could not identify from this PR: https://salsa.debian.org/multimedia-team/kodi-media-center/kodi/-/merge_requests/10

AFAIK drm prime can be added to the GL builds. The direct-to-plane method is only available on GBM. Wayland and x11 would have to use the EGL method.

**Comment by stsichler** [comment]
> @lrusak GL does not work with HW acceleration on NanoPi-R5C 
It's the same problem for me on my RPi4B and currently the only reason for me to rebuild Kodi from Sources instead of using the default armhf Debian package.

**Comment by amazingfate** [comment]
> > @lrusak GL does not work with HW acceleration on NanoPi-R5C and the platform I could not identify from this PR: https://salsa.debian.org/multimedia-team/kodi-media-center/kodi/-/merge_requests/10
> 
> AFAIK drm prime can be added to the GL builds. The direct-to-plane method is only available on GBM. Wayland and x11 would have to use the EGL method.

If kodi is built wih GL, in gbm session there is  also no drm prime setting. If direct-to-plane mode does't depend on EGL it should be possible to enable it even in a GL build.
Arm platforms are all having better GLES support than GL. Although some opensource drivers like panfrost and freedreno do support OpenGL, but they can't enable drm prime. I don't know much about x86 platform drm prime, do x86 platforms use GL for drm prime?

**Pull request merged**